[Getting started][Feature flags] Add "add your code" page (#6502)
PostHog/posthog.com: contents/docs/integrate/feature-flags-code/_snippets/feature-flags-code-react.mdx

There are two ways to implement feature flags in React:

1. Using hooks.
2. Using the `<PostHogFeature>` component.

> ### Method 1: Using hooks

Reviewer: ```suggestion
## Method 1: Using hooks
```

PostHog provides several hooks to make it easy to use feature flags in your React app.

| Hook | Description |
| --- | --- |
| `useFeatureFlagEnabled` | Returns a boolean indicating whether the feature flag is enabled. |
| `useFeatureFlagPayload` | Returns the payload of the feature flag. |
| `useFeatureFlagVariantKey` | Returns the variant key of the feature flag. |
| `useActiveFeatureFlags` | Returns an array of active feature flags. |

> #### Example 1: Using a boolean feature flag

Reviewer: ```suggestion
### Example 1: Using a boolean feature flag
```

```react
import { useFeatureFlagEnabled } from 'posthog-js/react'

function App() {
  const showWelcomeMessage = useFeatureFlagEnabled('flag-key')

  return (
    <div className="App">
      {
        showWelcomeMessage ? (
          <div>
            <h1>Welcome!</h1>
            <p>Thanks for trying out our feature flags.</p>
          </div>
        ) : (
          <div>
            <h2>No welcome message</h2>
            <p>Because the feature flag evaluated to false.</p>
          </div>
        )
      }
    </div>
  );
}

export default App;
```

> #### Example 2: Using a multivariate feature flag

Reviewer: ```suggestion
### Example 2: Using a multivariate feature flag
```

```react
import { useFeatureFlagVariantKey } from 'posthog-js/react'
import { useState, useEffect } from 'react'

function App() {
  const variantKey = useFeatureFlagVariantKey('show-welcome-message')
  const [ welcomeMessage, setWelcomeMessage ] = useState('')

  useEffect(() => {
    if (variantKey === 'variant-a') {
      setWelcomeMessage('Welcome to the Alpha!')
    } else if (variantKey === 'variant-b') {
      setWelcomeMessage('Welcome to the Beta!')
    }
  }, [variantKey])

  return (
    <div className="App">
      {
        welcomeMessage ? (
          <div>
            <h1>{welcomeMessage}</h1>
            <p>Thanks for trying out our feature flags.</p>
          </div>
        ) : (
          <div>
            <h2>No welcome message</h2>
            <p>Because the feature flag evaluated to false.</p>
          </div>
        )
      }
    </div>
  );
}

export default App;
```

> #### Example 3: Using a flag payload

Reviewer: ```suggestion
### Example 3: Using a flag payload
```

```react
import { useFeatureFlagPayload } from 'posthog-js/react'

function App() {
  const payload = useFeatureFlagPayload('show-welcome-message')

    return (
                <>
                {
                    payload?.welcomeMessage ? (
                        <div className="welcome-message">
                            <h2>{payload?.welcomeTitle}</h2>
                            <p>{payload.welcomeMessage}</p>
                        </div>
                    ) : <div>
                        <h2>No welcome message</h2>
                        <p>Because the feature flag evaluated to false.</p>
                    </div>
                }
        </>
    )
}
```

> ### Method 2: Using the PostHogFeature component

Reviewer: ```suggestion
## Method 2: Using the PostHogFeature component
```

> The PostHogFeature simplifies code by handling feature flag related logic. 

Reviewer: ```suggestion
The `PostHogFeature` component simplifies code by handling feature flag related logic. 
```

> It also has the added benefit of capturing metrics, like how many times a user interacts with this feature automatically.

Reviewer: ```suggestion
It also automatically captures metrics, like how many times a user interacts with this feature.
```

> **Note:** You still need the [`PostHogProvider`](/docs/libraries/react#installation) at the top level for this to work.

Here is an example:

```react
import { PostHogFeature } from 'posthog-js/react'

function App() {

    return (
        <PostHogFeature flag='show-welcome-message' match={true}>
            <div>
                <h1>Hello</h1>
                <p>Thanks for trying out our feature flags.</p>
            </div>
        </PostHogFeature>
    )
}
```

- The `match` on the component can be either `true`, or the variant key, to match on a specific variant.

- If you also want to show a default message, you can pass these in the `fallback` attribute.

If you wish to customise logic around when the component is considered visible, you can pass in `visibilityObserverOptions` to the feature. These take the same options as the [IntersectionObserver API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). By default, we use a threshold of 0.1. 

> #### Payloads

Reviewer: ```suggestion
### Payloads
```
